Data: Build the basic data controls into every store

jsnajdr opened the pull request:
In the `@wordpress/block-editor` package, replaces the home-grown `SELECT` control with `SYNC_SELECT` from `@wordpress/data-controls`.

There is one thing in this PR I don't like very much: the `@wordpress/block-editor` becomes indirectly dependent on `@wordpress/api-fetch`, although it never uses that code. The `controls` object exported from `@wordpress/data-controls` contains the `API_FETCH` field. Although we don't use it, it's there, and it's not even possible to tree-shake it away. It's a field of an exported object, not a standalone export.

How bad is that?

youknowriad commented:
i'm not sure about this PR tbh, I think the "syncSelect" control is small enough that duplication is probably better than having the ap-fetch dependency there.

jsnajdr commented:
> i'm not sure about this PR tbh, I think the "syncSelect" control is small enough that duplication is probably better than having the ap-fetch dependency there.

I tend to agree that the pros don't outweight the cons here. It shows, though, that there might be something slightly wrong with the `data-controls` package. Maybe it's putting together things that don't belong together?

The select/dispatch controls are so close to the data stores so that they could be part of the `data` package itself. And I would go even further: make them built-in in each store by default 🙂 Just like `rungen` has plenty of builtin controls like `all` or `fork` and then the user can register additional ones.

youknowriad commented:
> I tend to agree that the pros don't outweight the cons here. It shows, though, that there might be something slightly wrong with the data-controls package. Maybe it's putting together things that don't belong together?

> The select/dispatch controls are so close to the data stores so that they could be part of the data package itself. And I would go even further: make them built-in in each store by default

I agree here 💯 

cc @nerrad

nerrad commented:
Sounds very similar to the evolution of resolvers moving into the wp.data package. I'm not opposed to this being built-in to wp.data by default. Built-in controls would also get rid of the gotcha folks encounter with not defining a controls object when registering a store and that resulting in resolvers not working.

Would we deprecate the `data-controls` package all-together?

jsnajdr commented:
> Would we deprecate the `data-controls` package all-together?

That's one of the options: it would be deprecated, and implemented as a thin wrapper around the builtin `data` controls.

Of course, we'd need to find a new home for the `apiFetch` control. Maybe the `api-fetch` package itself? The control code is self-contained without any dependencies: a control creator that returns a `{ type: 'API_FETCH' }` object and a handler that executes `apiFetch`.

We could also fix the unfortunate naming of the `select` control discussed in #25235, namely making the `select` control do the sync select rather than `resolveSelect`.

There's one thing that's not great about moving the controls to the `@wordpress/data` package: how are we going to export them? The `select` and `dispatch` exports are already taken by the `defaultRegisty.select` and `defaultRegistry.dispatch` bindings. We'd need to find some less-than-optimal names for them.

This is where the `@wordpress/data-controls` package provides value: it makes the imports ergonomic:
```js
// import controls with nice names from dedicated package
import { select, dispatch } from '@wordpress/data-controls';

// and use them in my control
export function *myControl() {
  yield select( 'core', 'foo' );
  yield dispatch( 'core', bar() );
}
```

youknowriad commented:
Maybe we can export theme as "controls" variable, it's not tree-shakable but we have just three right?

jsnajdr committed:
createRegistrySelector: don't bind to defaultRegistry by default

We intend to register some controls created by `createRegistryControl` as
built-ins with every store. In order to do that, we need to break a dependency
cycle where registry creation depends on store registration which depends
on creating controls which depends on default registry.

The solution is to remove the `defaultRegistry` binding, which is there
only to satisfy a typechecker anyway, and doesn't have any runtime impact.

jsnajdr committed:
Data: add built-in controls for select, resolveSelect and dispatch

Instead of shipping them in a separate `data-controls` package and require
the store author to register them, build them in into every store.

jsnajdr committed:
data-controls: Remove implementation of built-in data controls

The implementation and the unit test have been moved to the `@wordpress/data`
package. The `data-controls` package now exposes only legacy aliases.

jsnajdr committed:
Update unit tests that depend on internal representation of data controls

Updates unit tests of actions (in `block-directory`, `edit-site` and `core-data`)
that inspect the internal properties (like `type`) of controls that the action
generator yields.

A test that verifies if the action _behaves_ correctly wouldn't need to be changed
like that.

jsnajdr committed:
Block Editor: replace a home-grown SELECT control with builtin one

jsnajdr commented:
I completely reworked this PR by moving the data controls (`select`, `resolveSelect`, `dispatch`) into the base `data` package and registering them by default with every store.

I also wrote a brand new unit test for the controls. I didn't like the original `data-controls` unit tests. In my view, they focused too much on internals (like working with `selector.hasResolver` flags) and were too far from how a real-world usage would look like. The new tests register some small example stores with controls, and verify the desired behavior by dispatching actions and calling selectors and checking if the results are as expected. There is zero mocking of anything.

The `data-controls` package now only exports aliases for backward compatibility.

The last commit migrates the `block-editor` package, just like the original PR, but this time without the unwanted dependency on `api-fetch`.

youknowriad commented on packages/data-controls/src/index.js:
```diff
@@ -31,98 +31,22 @@ export const apiFetch = ( request ) => {
-		args,
-	};
-}
+export const dispatch = dataControls.dispatch;
```
do you think we should deprecate these?

jsnajdr replied:
Yes, deprecation would discourage folks from adding new usages of `data-controls`. I guess the `data-controls` wrappers should use the `@wordpress/deprecated` helper?

jsnajdr replied:
@youknowriad I updated the `data-controls` wrappers to warn about deprecation using the `@wordpress/deprecated` package, in https://github.com/WordPress/gutenberg/pull/25362/commits/51f7fc52d7bcbe052a84b2352a02ea326c504f70.

But now I have problems with unit tests that I don't know how to solve: every unit tests uses the `jest-console` plugin that checks that the code didn't produce any console messages. But the `deprecated()` function logs the deprecation messages with `console.warn()`, so many unit tests for code that still uses the deprecated controls now fail.

Is there any established way how to get out of this?

youknowriad replied:
Yes, we have some asserters to aknowledge that warnings will be thrown. `expect( console ).toHaveWarned();`

That said, when we introduce deprecations, we do it after refactoring the Core usage of this API. Do we still have existing places where these APIs are used on Gutenberg?

jsnajdr replied:
> Do we still have existing places where these APIs are used on Gutenberg?

Yes, the `data-controls` are still used in many stores. It will take some 5+ PRs to migrate them one by one.

At this moment, it seems to me the best option is to mark the `data-controls` as deprecated only after that migration is finished. It shouldn't take long after this PR is merged.

youknowriad replied:
> it seems to me the best option is to mark the data-controls as deprecated only after that migration is finished. It shouldn't take long after this PR is merged.

👍

jsnajdr commented on packages/data/src/factory.js:
```diff
@@ -58,15 +51,6 @@ export function createRegistrySelector( registrySelector ) {
-	 *
-	 * @type {WPDataRegistry}
-	 */
-	selector.registry = defaultRegistry;
```
I described the motivation for this in the commit's commit message:

> We intend to register some controls created by `createRegistryControl` as built-ins with every store. In order to do that, we need to break a dependency cycle where registry creation depends on store registration which depends on creating controls which depends on default registry.
>
> The solution is to remove the `defaultRegistry` binding, which is there only to satisfy a typechecker anyway, and doesn't have any runtime impact.

Another solution would be to break up the `factory.js` module and put `createRegistrySelector` and `createRegistryControl` into separate modules. `createRegistrySelector` has the offending `defaultRegistry` dependency, while `createRegistryControl` is needed to create a store and therefore introduces a cyclic dependency.

However, to make the API really type-safe, we'd need to rethink how we create the selectors and introduce types like `UnregisteredRegistrySelector` and `RegisteredRegistrySelector`. Right now, the usage is like:
```js
const selector = createRegistrySelector( ( select ) => ( state ) => { select( 'core' ) /* ... */ } );

// at this moment the selector is in a weird state where it's either unbound or bound to `defaultRegistry`
// it's not entirely certain whether the following call should crash or not and what `value` is supposed to be
const value = selector( state ); 

// only after this registration is the `selector` registered and has 100% defined behavior
customRegistry.registerStore( 'mystore', { reducer, selectors:  { selector } } );

// this second call can have an entirely different result from the first one, although the `selector`
// value is still the same and has the same type. The `selector` properties have been mutated during
// `registerStore`.
const value = selector( state );
```
I.e., the code is not safe and I don't immediately see how introducing types could prevent the unsafe usages.

youknowriad replied:
I'm not sure i understand everything here but I agree that depending on the default registry should be avoided anyway. The default registry only exists because it was there first. If we'd start over, we'd probably avoid it entirely.

jsnajdr replied:
> I'm not sure i understand everything here but I agree that depending on the default registry should be avoided anyway.

The tl;dr is that the default registry reference started introducing dependency cycles 🙂

jsnajdr committed:
Mark the deprecated controls as deprecated

jsnajdr committed:
Comment out the deprecations until all Gutenberg packages are updated

youknowriad reviewed (approved):
LGTM 👍

jsnajdr commented:
Thanks for the review @youknowriad :+1: I'll check if unit tests are indeed green after the deprecations have been commented out, and then will merge the PR.
